kythe/typescript/indexer.ts

```diff
@@ -106,8 +122,83 @@ enum Context {
   Setter,
 }
 
+/** IndexerHost provides useful methods for indexing a TypeScript program. */
+class IndexerHost implements PluginHost {
+  /**
+   * rootDirs is the list of rootDirs in the compiler options, sorted
+   * longest first.  See this.moduleName().
+   */
+  rootDirs: string[];
+
+  /** A shorter name for the rootDir in the CompilerOptions. */
+  sourceRoot: string;
+
+  constructor(
+      /**
+       * The VName for the CompilationUnit, containing compilation-wide info.
+       */
+      private compilationUnit: VName,
+      /**
+       * A map of path to path-specific VName.
+       */
+      private pathVNames: Map<string, VName>,
+      /** All source file paths in the TypeScript program. */
+      public paths: string[],
+      public program: ts.Program,
+  ) {
+    this.sourceRoot = program.getCompilerOptions().rootDir || process.cwd();
+    let rootDirs = program.getCompilerOptions().rootDirs || [this.sourceRoot];
+    rootDirs = rootDirs.map(d => d + '/');
+    rootDirs.sort((a, b) => b.length - a.length);
+    this.rootDirs = rootDirs;
+  }
+
+  /**
+   * moduleName returns the ES6 module name of a path to a source file.
+   * E.g. foo/bar.ts and foo/bar.d.ts both have the same module name,
+   * 'foo/bar', and rootDirs (like bazel-bin/) are eliminated.
+   * See README.md for a discussion of this.
+   */
+  moduleName(sourcePath: string): string {
+    // Compute sourcePath as relative to one of the rootDirs.
+    // This canonicalizes e.g. bazel-bin/foo to just foo.
+    // Note that this.rootDirs is sorted longest first, so we'll use the
+    // longest match.
+    for (const rootDir of this.rootDirs) {
+      if (sourcePath.startsWith(rootDir)) {
+        sourcePath = path.relative(rootDir, sourcePath);
+        break;
+      }
+    }
+    return stripExtension(sourcePath);
+  }
+
+  /**
+   * pathToVName returns the VName for a given file path.
+   */
+  pathToVName(path: string): VName {
+    const vname = this.pathVNames.get(path);
+    return {
+      signature: '',
+      language: '',
+      corpus: vname && vname.corpus ? vname.corpus :
+                                      this.compilationUnit.corpus,
+      root: vname && vname.corpus ? vname.root : this.compilationUnit.root,
+      path: vname && vname.path ? vname.path : path,
+    };
+  }
+
+  /**
+   * emit emits a Kythe entry, structured as a JSON object.  Defaults to
+   * emitting to stdout but users may replace it.
+   */
+  emit = (obj: {}) => {
+    console.log(JSON.stringify(obj));
+  };
+}
+
 /** Visitor manages the indexing process for a single TypeScript SourceFile. */
-class Visitor {
+class Visitor extends IndexerHost {
```

Comment:
I think I'd just leave all the methods in Visitor, there's no real design benefit to separating the two.  It's unclear which properties/methods belong in which, and as the plugin needs more stuff we'd need to move them around.

Reply:
> I think I'd just leave all the methods in Visitor, there's no real design benefit to separating the two.

I tried this, but passing the Visitor as the `PluginHost` felt a little bit hacky because the Visitor is constructed for only one source file to be indexed:

https://github.com/kythe/kythe/blob/c2e3491259dabe1fb3c5aa9dfe55efc7c58893bc/kythe/typescript/indexer.ts#L153

which means that passing a Visitor to the plugins, we would have to construct a dummy Visitor to some arbitrary TypeScript file. Because this arbitrary file is never used in the `PluginHost`, I find it a little confusing and inelegant to include it - a conscious reader may question what `files[0]` in the Visitor construction `new Visitor(vname, pathVNames, paths, program, files[0])`, and while we can add a comment that the file is irrelevant, this seems less elegant to me than decoupling the Host from the actual visitor.

Please let me know if you disagree, I want to discuss this.

Reply:
Nope, your reasoning makes sense!

One other idea: would it make sense for Visitor to accept an IndexerHost as a parameter too?  That would mean we create one IndexerHost and pass it to both the Visitor and the Plugin.  In particular I wonder if there's a more meaningful name we can give IndexerHost, because right now it's basically "grab bag of all the stuff that Visitor and Plugin both need access to".

This LGTM as is if you don't have any better ideas, just thinking aloud.

Reply:
I'm wondering about the name "...Host". I wonder if the more generic "Context" might be better, e.g. PluginContext, IndexerContext, etc.

Reply:
How about `IndexerContext`? In this case, though, it would probably be best to get rid of the interface or rename it to something even more generic, which I'm not sure there is.

Reply:
Changed the interface to `IndexerHost` and the implementation to `IndexerContext`. The Visitor and plugins now both consume a common `IndexerContext`.